File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/compile/EqualsHashCodeGenerator.java
package me.bristermitten.mittenlib.annotations.compile;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import me.bristermitten.mittenlib.annotations.ast.Property;

import javax.inject.Inject;
import javax.lang.model.element.Modifier;
import java.util.List;
import java.util.Objects;

/**
 * Generates equals and hashCode methods for configuration classes.
 * This class creates standard implementations that compare all properties
 * of a configuration class for equality and generate consistent hash codes.
 */
public class EqualsHashCodeGenerator {
    private final MethodNames methodNames;

    @Inject
    public EqualsHashCodeGenerator(MethodNames methodNames) {
        this.methodNames = methodNames;
    }

    /**
     * Generates an equals method for a configuration class.
     * The generated method follows the standard equals contract:
     * - It's reflexive: an object is equal to itself
     * - It handles null and different class types
     * - It compares all properties for equality using Objects.equals
     *
     * @param configClassName The name of the class for which the equals method is being generated
     * @param properties      The list of properties to compare in the equals method
     * @return A MethodSpec representing the generated equals method
     */
    public MethodSpec generateEquals(ClassName configClassName, List<Property> properties) {
        var builder = MethodSpec.methodBuilder("equals")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(boolean.class)
                .addParameter(Object.class, "o");

        builder.beginControlFlow("if (this == o)")
                .addStatement("return true")
                .endControlFlow();

        builder.beginControlFlow("if (o == null || getClass() != o.getClass())")
                .addStatement("return false")
                .endControlFlow();

        builder.addStatement("$T that = ($T) o", configClassName, configClassName);

        for (Property property : properties) {
            builder.addStatement("if (!$T.equals(this.$L(), that.$L())) return false",
                    Objects.class,
                    methodNames.safeMethodName(property),
                    methodNames.safeMethodName(property));
        }

        builder.addStatement("return true");

        return builder.build();
    }

    /**
     * Generates a hashCode method for a configuration class.
     * The generated method uses Objects.hash to create a hash code based on all properties,
     * ensuring that objects that are equal according to equals() will have the same hash code.
     *
     * @param properties The list of properties to include in the hash code calculation
     * @return A MethodSpec representing the generated hashCode method
     */
    public MethodSpec generateHashCode(List<Property> properties) {
        var builder = MethodSpec.methodBuilder("hashCode")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(int.class);

        builder.addStatement("return $T.hash($L)", Objects.class,
                properties.stream()
                        .map(property -> "this." + methodNames.safeMethodName(property) + "()")
                        .reduce((a, b) -> a + ", " + b)
                        .orElse(""));

        return builder.build();
    }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/compile/EqualsHashCodeGenerator.scala
package me.bristermitten.mittenlib.annotations.compile

import com.google.inject.Inject
import com.palantir.javapoet.{ArrayTypeName, ClassName, MethodSpec, TypeName}

import java.util.{List as JList, Objects as JObjects}
import javax.lang.model.element.Modifier
import me.bristermitten.mittenlib.annotations.ast.Property
import me.bristermitten.mittenlib.annotations.domain.{
  Property => DomainProperty
}

import scala.jdk.CollectionConverters.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.BlockBuilder.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.given

/** Generates equals and hashCode methods for configuration classes. This class
  * creates standard implementations that compare all properties of a
  * configuration class for equality and generate consistent hash codes.
  */
class EqualsHashCodeGenerator @Inject() (private val methodNames: MethodNames):

  private def toSharedFields(properties: JList[Property]): List[SharedField] =
    properties.asScala.toList.map { p =>
      val t = TypeRef.of(TypeName.get(p.propertyType()))
      val isArr = TypeName.get(p.propertyType()).isInstanceOf[ArrayTypeName]
      SharedField(
        name = p.name(),
        tpe = t,
        accessor = receiver => receiver.call(methodNames.safeMethodName(p)),
        isArray = isArr
      )
    }

  private def toSharedFieldsDomain(
      properties: JList[DomainProperty]
  ): List[SharedField] =
    properties.asScala.toList.map { p =>
      val t = TypeRef.of(TypeName.get(p.typeMirror))
      val isArr = TypeName.get(p.typeMirror).isInstanceOf[ArrayTypeName]
      SharedField(
        name = p.name,
        tpe = t,
        accessor = receiver => receiver.call(methodNames.safeMethodName(p)),
        isArray = isArr
      )
    }

  /** Generates an equals method for a configuration class.
    */
  def generateEquals(
      configClassName: ClassName,
      properties: JList[Property]
  ): MethodSpec =
    val methodDecl =
      BoilerplateHelper.equalsDecl(configClassName, toSharedFields(properties))
    CodeBlockRenderer.renderMethod(methodDecl)

  /** Generates an equals method for a configuration class (domain.Property
    * variant).
    */
  def generateEqualsDomain(
      configClassName: ClassName,
      properties: JList[DomainProperty]
  ): MethodSpec =
    val methodDecl =
      BoilerplateHelper.equalsDecl(
        configClassName,
        toSharedFieldsDomain(properties)
      )
    CodeBlockRenderer.renderMethod(methodDecl)

  /** Generates a hashCode method for a configuration class.
    */
  def generateHashCode(properties: JList[Property]): MethodSpec =
    val methodDecl = BoilerplateHelper.hashCodeDecl(toSharedFields(properties))
    CodeBlockRenderer.renderMethod(methodDecl)

  /** Generates a hashCode method for a configuration class (domain.Property
    * variant).
    */
  def generateHashCodeDomain(properties: JList[DomainProperty]): MethodSpec =
    val methodDecl =
      BoilerplateHelper.hashCodeDecl(toSharedFieldsDomain(properties))
    CodeBlockRenderer.renderMethod(methodDecl)
